### PR TITLE
Feature #168713 feat: Make Tj-certificate multi-agency aware

### PR DESCRIPTION
**Pull request by tushar33**

### CODE REVIEW

File: src/components/com_tjcertificate/administrator/controllers/agency.json.php [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+	/**
+	 * Function to check user agency
+	 *
+	 * @param   int  $agencyId  agency id
+	 * 
+	 * @return  integer  The integer of the primary key
+	 *
+	 * @since  __DEPLOY_VERSION__
+	 */
+	private function checkUserAgency($agencyId)
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/controllers/agency.json.php]: why do we need this function?

File: src/components/com_tjcertificate/administrator/models/certificates.php [hunk 2]
```diff
@@ -81,18 +88,64 @@ protected function getListQuery()
...
+		// Filter by client
+		$client = $this->getState('filter.client');
+
 		// Create the base select statement.
 		$query->select(array('ci.*', 'ct.title', 'users.name as uname'));
 		$query->from($db->quoteName('#__tj_certificate_issue', 'ci'));
+
 		$query->join('LEFT', $db->quoteName('#__tj_certificate_templates', 'ct') .
 			' ON (' . $db->quoteName('ci.certificate_template_id') . ' = ' . $db->quoteName('ct.id') . ')');
-		$query->join('LEFT', $db->quoteName('#__users', 'users') . ' ON (' . $db->quoteName('ci.user_id') . ' = ' . $db->quoteName('users.id') . ')');
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/models/certificates.php:95]: Why changed to inner join..we want to see certificates of deleted users as well

File: src/components/com_tjcertificate/administrator/models/certificates.php [hunk 3]
```diff
@@ -81,18 +88,64 @@ protected function getListQuery()
...
+				' AND ' . $db->qn('clusters.client') . " = 'com_multiagency' )");
+
+			$query->join('LEFT', $db->qn('#__tjmultiagency_multiagency', 'agency') .
+				' ON (' . $db->qn('agency.id') . ' = ' . $db->qn('clusters.client_id') . ')');
+
+			$agencyId = $this->getState('filter.agency_id');
+
+			// If don't have manage all user permission then get users of own agency
+			if (!$canManageAllAgencyUser && !$agencyId)
+			{
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/models/certificates.php:131]: put all $subquery related statement inside this IF statement

File: src/components/com_tjcertificate/administrator/models/certificates.php [hunk 4]
```diff
@@ -81,18 +88,64 @@ protected function getListQuery()
...
+				$query->where($db->quoteName('agency.id') . ' in (' . $subquery . ')');
+			}
+
+			if ($agencyId)
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/models/certificates.php]: use elseif

File: src/components/com_tjcertificate/administrator/models/certificates.php [hunk 5]
```diff
@@ -81,18 +88,64 @@ protected function getListQuery()
...
 		$query->join('LEFT', $db->quoteName('#__tj_certificate_templates', 'ct') .
 			' ON (' . $db->quoteName('ci.certificate_template_id') . ' = ' . $db->quoteName('ct.id') . ')');
-		$query->join('LEFT', $db->quoteName('#__users', 'users') . ' ON (' . $db->quoteName('ci.user_id') . ' = ' . $db->quoteName('users.id') . ')');
+
+		$query->join('INNER', $db->quoteName('#__users', 'users') .
+			' ON (' . $db->quoteName('ci.user_id') . ' = ' . $db->quoteName('users.id') . ')');
+
+		if (ComponentHelper::isEnabled('com_multiagency') && $this->params->get('enable_multiagency'))
+		{
+			$canManageAllAgencyUser = $user->authorise('core.manage.all.agency.user', 'com_multiagency');
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/models/certificates.php]: create a class var to store `com_multiagency` in every file you used.

File: src/components/com_tjcertificate/administrator/models/fields/allusers.php [hunk 6]
```diff
@@ -0,0 +1,94 @@
+<?php
+/**
+ * @version    SVN: <svn_id>
+ * @package    Com_Tjlms
+ * @author     Techjoomla <extensions@techjoomla.com>
+ * @copyright  Copyright (c) 2009-2015 TechJoomla. All rights reserved.
+ * @license    GNU General Public License version 2 or later.
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/models/fields/allusers.php]: Update comments

File: src/components/com_tjcertificate/administrator/models/fields/allusers.php [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+use Joomla\CMS\Factory;
+use Joomla\CMS\Language\Text;
+use Joomla\CMS\HTML\HTMLHelper;
+use Joomla\CMS\Component\ComponentHelper;
+
+JFormHelper::loadFieldClass('list');
+
+/**
+ * Supports an HTML select list of courses
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/models/fields/allusers.php]: update comment everywhere

File: src/components/com_tjcertificate/administrator/models/fields/allusers.php [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+	{
+		$db = Factory::getDbo();
+		$query = $db->getQuery(true);
+
+		// Select the required fields from the table.
+		$query->select('u.id, u.name, u.username');
+		$query->from('`#__users` AS u');
+		$query->where($db->qn('u.block') . ' = 0');
+		$query->order($db->escape('u.name' . ' ' . 'asc'));
+		$db->setQuery($query);
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/models/fields/allusers.php]: you can use users model here

File: src/components/com_tjcertificate/administrator/models/forms/filter_certificates.xml [hunk 9]
```diff
@@ -41,6 +41,14 @@
 			<option value="">COM_TJCERTIFICATE_CERTIFICATE_FILTER_ISSUED_USER_SELECT</option>
 		</field> -->
 
+		<field name="agency_id"
+			type="agencies"
+			filter="safehtml"
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/models/forms/filter_certificates.xml]: the filter is not correct...it should be integer...check and fix

File: src/components/com_tjcertificate/administrator/models/forms/trainingrecord.xml [hunk 10]
```diff
@@ -3,12 +3,21 @@
 	<fieldset>
 		<field name="id" type="hidden"/>
 
+		<field 
+			name="agency_id" 
+			type="agencies"
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/models/forms/trainingrecord.xml:8]: add filter attribute here

File: src/components/com_tjcertificate/administrator/models/trainingrecord.php [hunk 11]
```diff
@@ -77,7 +77,13 @@ public function getItem($pk = null)
 	public function getForm($data = array(), $loadData = true)
 	{
 		// Get the form.
-		$form = $this->loadForm('com_tjcertificate.trainingrecord', 'trainingrecord', array('control' => 'jform', 'load_data' => $loadData));
+		$form                 = $this->loadForm('com_tjcertificate.trainingrecord', 'trainingrecord', array('control' => 'jform', 'load_data' => $loadData));
+		$loggedInuser         = Factory::getUser();
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/models/trainingrecord.php]: delete extra spacings after var

File: src/components/com_tjcertificate/administrator/views/certificates/tmpl/default.php [hunk 12]
```diff
@@ -278,3 +291,22 @@
...
+		jQuery('#filter_agency_id').closest("div").remove();
+	});
+}
+else
+{
+	// Hide all users drop-down if multi-agency is enable
+	jQuery(document).ready(function() {
+		jQuery('#filter_user_id').closest("div").remove();
+	});
+}
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/views/certificates/tmpl/default.php]: I think this should be handled with view.html.php

File: src/components/com_tjcertificate/administrator/views/trainingrecord/tmpl/edit.php [hunk 13]
```diff
@@ -108,4 +118,12 @@ class="p-5"
 <script type="text/javascript">
 var allowedAttachments = '<?php echo $this->allowedFileExtensions; ?>';
 var attachmentMaxSize  = '<?php echo $this->uploadLimit; ?>';
+
+jQuery(document).ready(function() {
+    if (jQuery('#jform_agency_id').val())
+    {
+      jQuery('#jform_agency_id').trigger("change");
+    }
+});
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/views/trainingrecord/tmpl/edit.php]: why do we need this code?

File: src/components/com_tjcertificate/site/models/forms/filter_certificates.xml [hunk 14]
```diff
@@ -19,6 +19,14 @@
 			clientByUser="1"
 			onchange="this.form.submit();" />
 
+		<field name="agency_id"
+			type="agencies"
+			filter="safehtml"
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/site/models/forms/filter_certificates.xml]: use proper filter

File: src/components/com_tjcertificate/site/models/forms/trainingrecord.xml [hunk 15]
```diff
@@ -14,16 +14,22 @@
 			name="user_id"
 			type="hidden" />
 
+		<field 
+			name="agency_id" 
+			type="agencies"
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/site/models/forms/trainingrecord.xml:19]: use filter

File: src/components/com_tjcertificate/site/views/certificates/tmpl/my.php [hunk 16]
```diff
@@ -242,3 +258,24 @@
...
+	});
+}
+else
+{
+	// Hide all users drop-down if multi-agency is enable
+	jQuery(document).ready(function() {
+		jQuery('#filter_user_id').closest("div").remove();
+	});
+}
+
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/site/views/certificates/tmpl/my.php]: same code

File: src/components/com_tjcertificate/site/views/trainingrecord/tmpl/edit.php [hunk 17]
```diff
@@ -113,6 +119,11 @@ class="p-5"
 var allowedAttachments = '<?php echo $this->allowedFileExtensions; ?>';
 var attachmentMaxSize  = '<?php echo $this->uploadLimit; ?>';
 
-
+jQuery(document).ready(function() {
+    if (jQuery('#jform_agency_id').val())
+    {
+      jQuery('#jform_agency_id').trigger("change");
+    }
+});
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/site/views/trainingrecord/tmpl/edit.php]: same code

File: src/components/com_tjcertificate/administrator/controllers/agency.json.php (hunk 1 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/controllers/agency.json.php]: removed not useful now

File: src/components/com_tjcertificate/administrator/models/certificates.php (hunk 2 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/models/certificates.php:95]: updated

File: src/components/com_tjcertificate/administrator/models/certificates.php (hunk 4 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/models/certificates.php]: done

File: src/components/com_tjcertificate/administrator/models/certificates.php (hunk 5 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/models/certificates.php]: done

File: src/components/com_tjcertificate/administrator/models/fields/allusers.php (hunk 6 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/models/fields/allusers.php]: done

File: src/components/com_tjcertificate/administrator/models/fields/allusers.php (hunk 7 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/models/fields/allusers.php]: done

File: src/components/com_tjcertificate/administrator/models/fields/allusers.php (hunk 8 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/models/fields/allusers.php]: done

File: src/components/com_tjcertificate/administrator/models/forms/filter_certificates.xml (hunk 9 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/models/forms/filter_certificates.xml]: done

File: src/components/com_tjcertificate/administrator/models/forms/trainingrecord.xml (hunk 10 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/models/forms/trainingrecord.xml:8]: done

File: src/components/com_tjcertificate/administrator/models/trainingrecord.php (hunk 11 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/models/trainingrecord.php]: done

File: src/components/com_tjcertificate/administrator/views/trainingrecord/tmpl/edit.php (hunk 13 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/views/trainingrecord/tmpl/edit.php]: removed

File: src/components/com_tjcertificate/site/models/forms/filter_certificates.xml (hunk 14 above)
Comment (tushar33) [reply src/components/com_tjcertificate/site/models/forms/filter_certificates.xml]: done

File: src/components/com_tjcertificate/site/models/forms/trainingrecord.xml (hunk 15 above)
Comment (tushar33) [reply src/components/com_tjcertificate/site/models/forms/trainingrecord.xml:19]: done

File: src/components/com_tjcertificate/site/views/certificates/tmpl/my.php (hunk 16 above)
Comment (tushar33) [reply src/components/com_tjcertificate/site/views/certificates/tmpl/my.php]: removed

File: src/components/com_tjcertificate/site/views/trainingrecord/tmpl/edit.php (hunk 17 above)
Comment (tushar33) [reply src/components/com_tjcertificate/site/views/trainingrecord/tmpl/edit.php]: removed

File: src/components/com_tjcertificate/administrator/controllers/agency.json.php (hunk 1 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/controllers/agency.json.php:123]: removed

File: src/components/com_tjcertificate/administrator/models/certificates.php (hunk 2 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/models/certificates.php:95]: updated

File: src/components/com_tjcertificate/administrator/views/certificates/tmpl/default.php (hunk 12 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/views/certificates/tmpl/default.php]: done

File: src/components/com_tjcertificate/administrator/models/certificates.php (hunk 3 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/models/certificates.php:131]: done

File: src/components/com_tjcertificate/administrator/controllers/agency.json.php [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+
+		// Get the current user id
+		$user = Factory::getuser();
+
+		if (!$user->id)
+		{
+			return false;
+		}
+
+		if (!$user->authorise('core.manage.own.agency.user', 'com_multiagency'))
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/controllers/agency.json.php]: use class var for com_multiagency every file

File: src/components/com_tjcertificate/administrator/controllers/agency.json.php [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+		$query = $db->getQuery(true);
+		$query->select('distinct(u.id), u.name');
+		$query->from($db->quoteName('#__users', 'u'));
+
+		if ($agencyId)
+		{
+			$query->join('INNER', '#__tj_cluster_nodes AS cn ON cn.user_id = u.id');
+			$query->join('INNER', $db->qn('#__tj_clusters', 'clusters') .
+					' ON (' . $db->qn('clusters.id') . ' = ' . $db->qn('cn.cluster_id') .
+					' AND ' . $db->qn('clusters.client') . " = 'com_multiagency' ) ");
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/controllers/agency.json.php]: and use that class var here and in all the places where you are using com_multiagency text

File: src/components/com_tjcertificate/administrator/models/certificates.php [hunk 20]
```diff
@@ -21,6 +23,8 @@
  */
 class TjCertificateModelCertificates extends ListModel
 {
+	protected $multiagency = 'com_multiagency';
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/models/certificates.php]: $comMultiAgency

File: src/components/com_tjcertificate/administrator/models/certificates.php [hunk 21]
```diff
@@ -81,18 +88,63 @@ protected function getListQuery()
...
 		$query->join('LEFT', $db->quoteName('#__tj_certificate_templates', 'ct') .
 			' ON (' . $db->quoteName('ci.certificate_template_id') . ' = ' . $db->quoteName('ct.id') . ')');
-		$query->join('LEFT', $db->quoteName('#__users', 'users') . ' ON (' . $db->quoteName('ci.user_id') . ' = ' . $db->quoteName('users.id') . ')');
+
+		$query->join('LEFT', $db->quoteName('#__users', 'users') .
+			' ON (' . $db->quoteName('ci.user_id') . ' = ' . $db->quoteName('users.id') . ')');
+
+		if (ComponentHelper::isEnabled($this->multiagency) && $this->params->get('enable_multiagency'))
+		{
+			$canManageAllAgencyUser = $user->authorise('core.manage.all.agency.user', $this->multiagency);
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/models/certificates.php]: write step-by-step code...use this statement just before it is being used...so that it is logically at the right place while debugging...put this before the line 129

File: src/components/com_tjcertificate/administrator/models/certificates.php [hunk 22]
```diff
@@ -194,6 +243,8 @@ protected function getListQuery()
 		$orderCol  = $this->state->get('list.ordering', 'ci.id');
 		$orderDirn = $this->state->get('list.direction', 'desc');
 
+		$query->group('ci.id');
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/models/certificates.php:246]: why is this needed?...this will affect the other certificate type which is added programmatically

File: src/components/com_tjcertificate/administrator/views/certificates/view.html.php [hunk 23]
```diff
@@ -116,6 +121,22 @@ public function display($tpl = null)
...
+
+		if ($this->isAgencyEnabled)
+		{
+			$this->filterForm->removeField('user_id', 'filter');
+		}
+		else
+		{
+			$this->filterForm->removeField('agency_id', 'filter');
+		}
+
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/views/certificates/view.html.php:135]: Check if the below code works

```php
if (ComponentHelper::isEnabled('com_multiagency') && $this->params->get('enable_multiagency'))
{
	$this->isAgencyEnabled = true;
        $this->filterForm->removeField('user_id', 'filter');
}
else
{
       $this->filterForm->removeField('agency_id', 'filter');
}

File: src/components/com_tjcertificate/site/views/certificates/view.html.php [hunk 24]
```diff
@@ -133,6 +144,15 @@ public function display($tpl = null)
...
 
+		if ($this->isAgencyEnabled)
+		{
+			$this->filterForm->removeField('user_id', 'filter');
+		}
+		else
+		{
+			$this->filterForm->removeField('agency_id', 'filter');
+		}
+
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/site/views/certificates/view.html.php:152]: try same code as suggested in the above files

Comment (tushar33) [reply src/components/com_tjcertificate/site/views/certificates/view.html.php:152]: done

File: src/components/com_tjcertificate/administrator/views/certificates/view.html.php (hunk 23 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/views/certificates/view.html.php:135]: done

File: src/components/com_tjcertificate/administrator/models/certificates.php (hunk 22 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/models/certificates.php:246]: showing repeated entry when user connected with multiple organizations so need to group by certificate id

File: src/components/com_tjcertificate/administrator/models/certificates.php (hunk 21 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/models/certificates.php]: done

File: src/components/com_tjcertificate/administrator/models/certificates.php (hunk 20 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/models/certificates.php]: done

File: src/components/com_tjcertificate/administrator/controllers/agency.json.php (hunk 19 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/controllers/agency.json.php]: done

File: src/components/com_tjcertificate/administrator/controllers/agency.json.php (hunk 18 above)
Comment (tushar33) [reply src/components/com_tjcertificate/administrator/controllers/agency.json.php]: done

File: src/components/com_tjcertificate/site/views/trainingrecord/tmpl/edit.php [hunk 25]
```diff
@@ -102,6 +107,7 @@ class="p-5"
 		</div>
 	</div>
 	
+	<input type="hidden" id="assigned_user_id" value="<?php echo $this->item->user_id; ?>" />
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/site/views/trainingrecord/tmpl/edit.php:110]: is this needed since you have already rendered the same field at line 55

File: src/components/com_tjcertificate/administrator/views/trainingrecord/tmpl/edit.php [hunk 26]
```diff
@@ -95,6 +104,7 @@ class="p-5"
 			<?php echo $this->form->renderField('comment'); ?>
 
 		</div>
+		<input type="hidden" id="assigned_user_id" value="<?php echo $this->item->user_id; ?>" />
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/views/trainingrecord/tmpl/edit.php:107]: same here. this field already rendered at line 67

Comment (tushar33) [reply src/components/com_tjcertificate/administrator/views/trainingrecord/tmpl/edit.php:107]: This is hidden field to show the user selected, in ajax we are using this value to show selected and Joomla field id is different from this

File: src/components/com_tjcertificate/site/views/trainingrecord/tmpl/edit.php (hunk 25 above)
Comment (tushar33) [reply src/components/com_tjcertificate/site/views/trainingrecord/tmpl/edit.php:110]: yes needed show the user selected, in ajax we are using this value to show selected

File: src/components/com_tjcertificate/administrator/models/trainingrecord.php [hunk 27]
```diff
@@ -285,6 +314,41 @@ public function validate($form, $data)
...
+					$assignedUserAgencyArr[] = $assignedUserAgency->id;
+				}
+
+				foreach ($loggedInUserAgencies as $loggedInUserAgency)
+				{
+					$loggedInUserAgencyArr[] = $loggedInUserAgency->id;
+				}
+
+				// Compare both users agencies
+				$result = array_intersect($loggedInUserAgencyArr, $assignedUserAgencyArr);
```
Comment (praneettekdi) [inline src/components/com_tjcertificate/administrator/models/trainingrecord.php:335]: can you explain this to me on call?

Comment (tushar33) [reply src/components/com_tjcertificate/administrator/models/trainingrecord.php:335]: As discussed added array_column
